File: src/decoders/news-question.ts
import type { SessionHandle } from "~/models";
import type { NewsQuestion } from "~/models/news-question";
import { decodeAttachment } from "./attachment";
import { decodePronoteDate } from "./pronote-date";
import { NewsQuestionKind } from "~/models/news-question-kind";
import { decodeNewsQuestionChoice } from "./news-question-choice";
import { decodeDomain } from "./domain";

export const decodeNewsQuestion = (question: any, session: SessionHandle): NewsQuestion => {
  const kind: NewsQuestionKind = question.genreReponse;
  const answered: boolean = question.reponse.V.avecReponse;

  let selectedAnswers: number[] | undefined;
  let textInputAnswer: string | undefined;

  if (answered && question.reponse.V.valeurReponse) {
    if (typeof question.reponse.V.valeurReponse === "string") {
      textInputAnswer = question.reponse.V.valeurReponse;
    }
    else {
      selectedAnswers = decodeDomain(question.reponse.V.valeurReponse.V);
      textInputAnswer = question.reponse.V.valeurReponseLibre;
    }
  }

  return {
    id: question.N,
    fullTitle: question.L,
    title: question.titre,
    position: question.rang,
    kind,

    maximumLength: question.tailleReponse,
    shouldRespectMaximumChoices: question.avecMaximum,
    maximumChoices: question.nombreReponsesMax,
    content: question.texte.V,

    attachments: question.listePiecesJointes.V.map((attachment: any) => decodeAttachment(attachment, session)),

    answerID: question.reponse.V.N,
    answered,

    shouldAnswer: question.reponse.V.estReponseAttendue,
    answerDate: question.reponse.V.reponduLe?.V && decodePronoteDate(question.reponse.V.reponduLe.V),

    choices: kind === NewsQuestionKind.TextInput ? [] : question.listeChoix.V.map(decodeNewsQuestionChoice),

    selectedAnswers,
    textInputAnswer
  };
};

File: library/src/commonMain/kotlin/decoders/news-question.kt
package decoders

import kotlinx.serialization.json.*
import models.NewsQuestion
import models.NewsQuestionKind
import models.SessionInformation

fun decodeNewsQuestion (question: JsonObject, sessionInformation: SessionInformation): NewsQuestion {
    val kind: NewsQuestionKind = NewsQuestionKind.fromInt(question["genreReponse"]!!.jsonPrimitive.int)
    val answered = question["reponse"]!!.jsonObject["V"]!!.jsonObject["avecReponse"]!!.jsonPrimitive.boolean

    var selectedAnswers: List<Int>? = null
    var textInputAnswer: String? = null

    if (answered && question["reponse"]!!.jsonObject["V"]!!.jsonObject["valeurReponse"] != null) {
        if (question["reponse"]!!.jsonObject["V"]!!.jsonObject["valeurReponse"]!!.jsonPrimitive.isString)
            textInputAnswer = question["reponse"]!!.jsonObject["V"]!!.jsonObject["valeurReponse"]!!.jsonPrimitive.content
        else {
            selectedAnswers = decodeDomain(question["reponse"]!!.jsonObject["V"]!!.jsonObject["valeurReponse"]!!.jsonObject["V"]!!.jsonPrimitive.content)
            textInputAnswer = question["reponse"]!!.jsonObject["V"]!!.jsonObject["valeurReponseLibre"]!!.jsonPrimitive.content
        }
    }

    return NewsQuestion(
        id = question["N"]!!.jsonPrimitive.content,
        fullTitle = question["L"]!!.jsonPrimitive.content,
        title = question["titre"]!!.jsonPrimitive.content,
        position = question["rang"]!!.jsonPrimitive.int,
        kind = kind,

        maximumLength = question["tailleReponse"]!!.jsonPrimitive.int,
        shouldRespectMaximumChoices = question["avecMaximum"]!!.jsonPrimitive.boolean,
        maximumChoices = question["nombreReponsesMax"]!!.jsonPrimitive.int,
        content = question["texte"]!!.jsonObject["V"]!!.jsonPrimitive.content,

        attachments = question["listePiecesJointes"]!!.jsonObject["V"]!!.jsonArray.map { decodeAttachment(it.jsonObject, sessionInformation) },

        answerId = question["reponse"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content,
        answered =  answered,

        shouldAnswer = question["reponse"]!!.jsonObject["V"]!!.jsonObject["estReponseAttendue"]!!.jsonPrimitive.boolean,
        answerDate = if (question["reponse"]!!.jsonObject["V"]!!.jsonObject["reponduLe"] != null) decodePronoteDate(question["reponse"]!!.jsonObject["V"]!!.jsonObject["reponduLe"]!!.jsonObject["V"]!!.jsonPrimitive.content) else null,

        choices = if (kind == NewsQuestionKind.TextInput) listOf() else question["listeChoix"]!!.jsonObject["V"]!!.jsonArray.map { decodeNewsQuestionChoice(it.jsonObject) },

        selectedAnswers = selectedAnswers,
        textInputAnswer = textInputAnswer
    )
}